front/app/utils/errorUtils.ts

```diff
@@ -91,10 +92,29 @@ export function getApiErrorMessage(
   inputTerm?: InputTerm,
   field?: string
 ) {
+  // We don't get the values (ie the min char count a text input, so they have to be hard-coded in the translations and kept in sync with the BE hardcoded value).
   return (
     messages[`api_error_${inputTerm}_${field}_${error}`] ||
     messages[`api_error_${field}_${error}`] ||
     messages[`api_error_${error}`] ||
     messages[`api_error_invalid`]
   );
 }
+
+export function getAjvErrorMessage(
+  error: ErrorObject['keyword'],
+  inputTerm?: InputTerm,
+  format?: string,
+  location?: string
+) {
+  console.log(error, inputTerm, format, location);
+  // Here the vales contained in ErrorMessage['params'] should be passed along for translation.
+  return (
+    messages[`ajv_error_${inputTerm}_${location}_${error}`] ||
```

Comment:
I'm not 100% following the whole error system, could help if you guide me through it for better feedback.

On first sight though, I'm a bit worried about mixing all specific error messages into one messages file, in one namespace.

Actually, sidetrack: It's a general thought popping up when I'm going through the code: Shouldn't we make our controls more opinionated for built-in fields?

I sometimes have a hard time following the controls, and I think it's caused by the fact that we're trying hard to render all our opinionated built-in fields with custom generic controls. I'm wondering whether it wouldn't be better to make more opinionated controls for all our built-in fields, that map 1:1 and are only used for those built-in fields. We can e.g. include `code` into the `options` of the UI schema and make the matcher detect it. It's accepting a bit of duplication, to gain way cleaner abstractions that will stand the test of time, instead of over time having generic controls being blown up with many options and matching conditionals to handle those almost-the-same-but-not-quite cases.

Back to the error copy: Then the opinionated custom copy related to that field can live within a messages file of the specific control, where it would make more sense than in the global namespace. The global namespace, or the generic custom fields, can define the generic error messages when we don't have more context.

Reply:
I agree with the general idea, I think the inputTerm being here although completely irrelevant to custom fields is a good example already of how this will go wrong. I consider this the one blocking point and will come up with a solution today. I see two paths : At the form level (passing in error messages overwrites to the form that will match errors on conrols with specific codes, or overwrite basic error messages with form context) or at the control level (fetching the custom control based on the code which will hold their own error messages). The second option is not the jsonforms way and will make it much more difficult to implement so I'll be going in the first direction, at least as far as errors go.